src/System.Linq/src/System/Linq/Concatenate.cs

```diff
@@ -45,5 +32,212 @@ private static IEnumerable<TSource> PrependIterator<TSource>(IEnumerable<TSource
+            private readonly IEnumerable<TSource> _third;
+
+            internal Concat3Iterator(IEnumerable<TSource> first, IEnumerable<TSource> second, IEnumerable<TSource> third)
+            {
+                Debug.Assert(first != null && second != null && third != null);
+                _first = first;
+                _second = second;
+                _third = third;
+            }
+
+            public override Iterator<TSource> Clone()
+            {
+                return new Concat3Iterator<TSource>(_first, _second, _third);
+            }
+
+            internal override ConcatIterator<TSource> Concat(IEnumerable<TSource> next)
+            {
+                return new ConcatNIterator<TSource>(this, next, 3);
+            }
+
+            internal override IEnumerable<TSource> GetEnumerable(int index)
+            {
+                switch (index)
+                {
+                    case 0: return _first;
+                    case 1: return _second;
+                    case 2: return _third;
+                    default: return null;
+                }
+            }
+        }
+
+        private sealed class ConcatNIterator<TSource> : ConcatIterator<TSource>
+        {
+            // To handle chains of >= 4 sources, we chain the concat iterators together and allow
+            // GetEnumerable to fetch enumerables from the previous sources.  This means that rather
+            // than each MoveNext/Current calls having to traverse all of the previous sources, we
+            // only have to traverse all of the previous sources once per chained enumerable.  An
+            // alternative would be to use an array to store all of the enumerables, but this has
+            // a much better memory profile and without much additional run-time cost.
+
+            private readonly ConcatIterator<TSource> _previousConcat;
+            private readonly IEnumerable<TSource> _next;
+            private readonly int _nextIndex;
+
+            internal ConcatNIterator(ConcatIterator<TSource> previousConcat, IEnumerable<TSource> next, int nextIndex)
+            {
+                Debug.Assert(previousConcat != null);
+                Debug.Assert(next != null);
+                Debug.Assert(nextIndex > 0);
+                _previousConcat = previousConcat;
+                _next = next;
+                _nextIndex = nextIndex;
+            }
+
+            public override Iterator<TSource> Clone()
+            {
+                return new ConcatNIterator<TSource>(_previousConcat, _next, _nextIndex);
+            }
+
+            internal override ConcatIterator<TSource> Concat(IEnumerable<TSource> next)
+            {
+                return new ConcatNIterator<TSource>(this, next, checked(_nextIndex + 1));
+            }
+
+            internal override IEnumerable<TSource> GetEnumerable(int index)
+            {
+                return
+                    index < _nextIndex ? _previousConcat.GetEnumerable(index) :
```

Comment:
If the case we're concerned about is a long list of enumerables, that means we're going to be allocating a large array to store them all.  That's an additional cost that's much more difficult to quantify in terms of whether this change is a win or not, whereas what's here with pencil and paper can be shown to be better than what's checked in.  It's also purely additive on top of this.  You ok then with deferring such a change until such a time when someone can do due diligence in analyzing its impact?  Any time there's talk of introducing a large array allocation where there wasn't one previously, I get a bit squeamish.